src/Nethermind/Nethermind.Consensus/GC/GCScheduler.cs

```diff
@@ -0,0 +1,192 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using System.Runtime;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Nethermind.Consensus;
+
+public sealed class GCScheduler
+{
+    // Thresholds and limits for triggering garbage collection
+    private const int BlocksBacklogTriggeringManualGC = 4;
+    private const int MaxBlocksWithoutGC = 250;
+    private const int MinSecondsBetweenForcedGC = 120;
+
+    // Flag indicating if a garbage collection is currently in progress
+    private static int _isPerformingGC = 0;
+
+    // Timer for scheduling periodic garbage collections when idle
+    private readonly Timer _gcTimer;
+    private readonly Stopwatch _stopwatch = new();
+    private Task _lastGcTask = Task.CompletedTask;
+    private bool _isNextGcBlocking = false;
+    private bool _isNextGcCompacting = false;
+    private bool _gcTimerSet = false;
+    private bool _fireGC = false;
+    private long _countToGC = 0L;
+
+    // Singleton instance of GCScheduler
+    public static GCScheduler Instance { get; } = new GCScheduler();
+
+    private GCScheduler()
+    {
+        // Initialize the timer without starting it
+        _gcTimer = new Timer(_ => PerformFullGC(), null, Timeout.Infinite, Timeout.Infinite);
+    }
+
+    /// <summary>
+    /// Activates background garbage collection when the processing queue is idle.
+    /// </summary>
+    /// <param name="queueCount">Number of items in the processing queue.</param>
+    public void SwitchOnBackgroundGC(int queueCount)
+    {
+        if (_fireGC)
+        {
+            _countToGC--;
+            // Trigger GC if block limit reached and minimum time elapsed
+            if (_countToGC <= 0 && _stopwatch.Elapsed.TotalSeconds > MinSecondsBetweenForcedGC)
+            {
+                _fireGC = false;
+                _stopwatch.Reset();
+                // Ensure only one GC task runs at a time
+                if (_lastGcTask.IsCompleted)
+                {
+                    _lastGcTask = PerformFullGCAsync();
+                }
+            }
+        }
+
+        // Avoid setting the timer if there are items in the queue or timer is already set
+        if (queueCount > 0 || _gcTimerSet)
+        {
+            // Don't switch on if still processing blocks
+            return;
+        }
+        _gcTimerSet = true;
+        // Schedule GC every 2 minutes when idle
+        _gcTimer.Change(TimeSpan.FromMinutes(2), TimeSpan.FromMinutes(2));
```

Comment:
Minor/me: I dislike `if return` pattern unless on top of the method, would prefer nesting.

Reply:
Inverted the ifs